config/v1/types_infrastructure.go

```diff
@@ -500,6 +505,58 @@ type AWSResourceTag struct {
 	Value string `json:"value"`
 }
 
+// DNSAndLBConfiguration contains information about API and API-Int Load Balancers
+// and who is responsible for that configuration.
+// +union
+type DNSAndLBConfiguration struct {
+	// provider determines which DNS solution is in use for this cluster.
+	// When the user has pre-created the Load Balancers for API and API-Int and the
+	// the DNS solution has already been configured, this field is set to "UserProvided".
+	// When the cluster is resposible for its own DNS configuration and LB creation,
+	// this field is set to "ClusterProvided".
+	// +default="ClusterProvided"
+	// +kubebuilder:default:="ClusterProvided"
+	// +kubebuilder:validation:Enum="UserProvided";"ClusterProvided"
+	// +unionDiscriminator
```

Comment:
It's not clear to me why we would use a diiscriminated union here. The mutually-exclusive fields in the union are the same type `LBConfigType`. Why not just have a single field for LBConfig which includes an enum field to distinguish cluster vs user-provided. The main benefit I see of a union discriminator is when you have mutually-exclusive fields of different types.

Reply:
When the union discriminator is of the default type "ClusterProvided", we don't need to save the LB names. But, in the case of "UserProvided", we do. So, the union represents this case best IMO.

Reply:
We don't need to save LB name in either case right? These values will not be used. 

To keep track of user-provided LBs, we should tag them to associate them with the cluster, similar to the way we do with user-provided subnets. Installer-provisioned LBs are also tagged, so this would be consistent across all clusters.

Unless we have a reason for these fields, I think it's better to remove them and defer the design of the API to when we have a need or reason to include them.

Reply:
> It's not clear to me why we would use a diiscriminated union here. 

Looking at [LoadBalancerType ](https://github.com/openshift/api/pull/1391)  a model for this API, I see that the PlatformLoadBalancer structs are unions but with only union discriminators--no member fields. So I'm assuming that the reason to use a union there is future proofing, which is fine with me and something I would leave up to the discretion of more knowledgeable api reviewers.

Reply:
In the on-prem BYO LoadBalancer scenario, the customer is completely responsible for managing the loadbalancers so they do not provide the name of the Load Balancers to the cluster to being with. 
In the custom DNS case, the user needs to provide the LB names so that the Installer can configure them. So, not exactly the same data is provided via install-config.
I am unclear about the best practices around this. If the goal is to accurately represent that state of the cluster, then the LB names need also make it into the Infrastructure Status. If the goal is to only add items that other components in the cluster read to take action, then the LB Names are not necessary as @patrickdillon pointed out.
I am Ok with adding the LB names at a later point but don't see it adding too much of an overhead now.  I'll let @JoelSpeed weigh in before I go ahead and remove the LB names.

Reply:
If the fields aren't going to be consumed, lets not add them, no point adding them until we have something that is a concrete consumer.

And yes, by having a union we have the option to expand this later and diverge from being the same across both. If we don't think that's going to be a future we see, flipping this round to have the type inside the same struct as the names isn't an unreasonable suggestion IMO

Reply:
Updated implementation to match discussion in this thread.